### ISSUE TITLE
File analysis runs indefinitely after upgrading to v2024.9.1

### ISSUE THREAD

**Issue by AlexanderPodorov** [microsoft/pyright#9146]
## Environment data

-   Pylance version: v2024.9.1
-   OS and version: MacOS 13.0.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.12.5

## Repro Steps

Upgrade to the specified version and file analysis runs indefinitely.

## Expected behavior

Upgrade to the specified version and file analysis should complete in a reasonable time.

## Actual behavior

Upgrade to the specified version and file analysis runs indefinitely.

## Logs
Repeated lines like that:
```
2024-09-12 13:59:15.147 [error] [
  `Failed to start language server, Class name = h, completed in 0ms, has a falsy return value, Arg 1: <Uri:/Users/..>, Arg 2:

[Error: LanguageStatusItem with id 'python.analysis.typeCheckingMode' ALREADY exists

2024-09-12 13:59:15.148 [error] [
  'Starting language server, Class name = d, completed in 1ms, has a falsy return value, , Return Value: undefined',
  [Error: LanguageStatusItem with id 'python.analysis.typeCheckingMode' ALREADY exists

2024-09-12 13:59:15.148 [error] [
  'Restarting language server, Class name = d, completed in 2ms, has a falsy return value, , Return Value: undefined',
  [Error: LanguageStatusItem with id 'python.analysis.typeCheckingMode' ALREADY exists
```

**Comment by erictraut** [comment]
Here's what's happening. When pyright encounters the `match` statement in your code sample, it attempts to narrow the type of the subject expression (`a, b`) based on the case statements. Initially, the type of `a, b` is `tuple[UA, UB]`. The first case statement matches against `tuple[A1, A2]`. In the fall-through case, the remaining type is `tuple[UA - A1, UB] | tuple[UA, UB - B1]` (where `-` denotes a subtraction from the original union type). You can see that this type is now a union with two subtypes. The fall-through case for the second case statement creates a union with four subtypes, etc. Each case statement doubles the number of subtypes in the resulting union. After `x` case statements, the resulting type will be a union with `2^x` entries. Creating new unions requires pyright to deduplicate subtypes, and this is an `O(n^2)` operation. To create a union with `2^x` entries, it becomes an `O(2^2x)` operation.

To address this, I've added a limit check. If the narrowing operation results in a union that has more than 128 entries, it falls back on a type that is less-accurate (not as narrow as it could be) but more manageable. I picked 128 empirically. If it's greater than this, the performance delay starts to become noticeable. With this limit in place, you won't see this hang, but type accuracy may suffer. For example, pyright won't be able to prove pattern matching exhaustion in some cases. Then again, in this use case you probably don't care about matching every conceivable combination of `UA` and `UB` subtypes with individual case statements.

This limit check will be included in the next release.

@AlexanderPodorov, thanks again for taking the time to help isolate a repro for this issue.

**Comment by AlexanderPodorov** [comment]
@erictraut , thanks a lot for finding the root cause of the issue!

> in this use case you probably don't care about matching every conceivable combination of UA and UB subtypes with individual case statements.

Yes, right. If `x = y`, then `Ax` is "compatible" with `Bx`, otherwise this is either an exception or returning `False`.
Like if we build a square matrix where indexes of `A` are columns and indexes of `B` are rows, diagonal elements are only valid combinations.

Can you suggest a better way to do it? I know it can be rewritten with a bunch of `if` and `isinstance` checks, but I figured `match` is less ugly way to achieve it. Thanks!

**Comment by erictraut** [comment]
One possible approach (assuming your classes are not intended to be subclassed) is to use a dictionary:

```python
allowed: dict[type[UA], type[UB]] = {
    A1: B1,
    A2: B2,
    A3: B3,
    ...
}

def test(a: UA, b: UB) -> bool:
    return allowed.get(type(a)) is type(b)
```

**Comment by AlexanderPodorov** [comment]
Thanks, I will try to figure this out.
I've oversimplified my original example. `dict` approach won't work for my use case, because instead of just `return True` I'm calling specific methods on those classes as well. `Ax` and `By` are "logically" compatible, but not from type checker point of view. Anyways, thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.384
